File: src/medium/347-top-k-frequent-elements/typescript/main.ts
// eslint-disable-next-line @typescript-eslint/naming-convention
function topKFrequent(nums: number[], k: number): number[] {
	const freq = new Map<number, number>();

	for (const number_ of nums) {
		freq.set(number_, (freq.get(number_) ?? 0) + 1);
	}

	const buckets = Array.from({length: nums.length + 1}).map(() => []);
	for (const [number_, count] of freq.entries()) {
		(buckets as number[][])[count].push(number_);
	}

	const result: number[] = [];
	for (let i = buckets.length - 1; i >= 0 && result.length < k; i--) {
		result.push(...(buckets as number[][])[i]);
	}

	return result;
}

File: src/medium/347-top-k-frequent-elements/python/main.py
from collections import Counter, defaultdict
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequency = Counter()
        for num in nums:
            frequency[num] += 1
        
        buckets = defaultdict(list)
        for [number_, count] in frequency.items():
            buckets[count].append(number_)

        result = []
        for count in range(len(nums), 0, -1):
            if count in buckets:
                result += buckets[count]
                if len(result) >= k:
                    break

        return result